bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/storage/url/BundleResourceHandler.java

```diff
@@ -94,16 +94,32 @@ protected void parseURL(URL url, String str, int start, int end) {
 				}
 			host = spec.substring(bundleIdIdx, bundleIdEnd);
 		}
-		if (pathIdx < end && spec.charAt(pathIdx) == '/')
-			path = spec.substring(pathIdx, end);
-		else if (end > pathIdx) {
+		// Extract query and fragment from the spec
+		// Query comes after '?' and before '#' (fragment)
+		String query = null;
+		int queryIdx = spec.indexOf('?', pathIdx);
+		int fragmentIdx = spec.indexOf('#', pathIdx);
+		int pathEnd = end;
+		
+		// Determine where the path ends (before query or fragment)
```

Comment:
Confirming the intention.  If the fragment `#` comes before the query `?` then you treat the `?` as part of the ref?